### PR TITLE
More features in prepare_report.py

### PR DESCRIPTION
**Pull request by cameel**
~Depends on #10679.~ Already on `develop`.

This PR adds a few nice-to-have but non-essential features `prepare_reprot.py`. We can live without them but they were pretty handy when debugging so maybe we want to keep them after all.
- Progress display - the script prints a dot or an error flag for every processed file.
- Summary - you can see how many errors there were. The exact number may not matter but 
- An option to stop on non-zero exit code from compiler and display the error (though not all that useful with Standard JSON which always reports success).
- Option to set the name of the report file.

The typical script output now looks more or less like this:
```
.......E.E..EEE.....EEE.E.E.E...EE.E.E.E...E....E.E.E.EE...EEE..EE...EE...EE.EE...E.EE.E.E..........EEEEE.E......E......EEEEEEEEE.E.EEEE.EE..E....EE.....E.EE.EEEE.....EE....E.E........E..........E...E
E..EE...EEEEE....EEE.E.EE.....E.EEEE.....EEE.E.E..E.EEE.E.EE.E....EE..E.E....EEEE.E.EE...E......E..E.E..E.EEE...EEEE....E.EE.E...E.......E..E....E..EE...E...E...E..EEEE...E.E..E.EEE....E.E...E...EE...
EEE.EEE.E.E.....EEEE...E......E.EEEE...E.......E..EEE..E..E....EE.E......E.E...EE.E...E...EE...EEEEE...E...E.E.E.EE.EE....E...EE......EE.E.EEE.EEEE..E.E.EEEEE....E...E.E......E...EE.E..E.....E.E....E.
EE.....E.E.E.EEE....E......E.E.E..EEE...E..E...E.E....E.E.....EE..EE.E..EEE.E.EE.E..........EE.E.E.E.E...E..EE..EE..E.EE.E.EEEEE..EE.EEE.EEE....E.E....E...E.EE.E.EEE.E.....E....E.EEE.E.E..EEEEEEE..E..
.EEE..E.EEE.E.E.EE....E.E.E.EEE.E...E..EE..EEEE.EEE.E...EE.EE..E..E......EEEEEE.EEEEE.....E..E..E..EE....EE..E.E......EEEE.EE..EE.EE.EEEE.EE.E.E.EEE..E...E.E..EEE....EEEEE...E..E....EE..EE...EEE....E.
E...EE.EE....E.EE.EEEE.E.E.E...E.E..EE..EEEEE.E......E...EE.EEE.EE.EEEEE..E..EEE..E.EE..E..E..EE.EE..E.E..E..E.EEE.EE.E..EEEEE..EE.E.EE..E.E.E..EE..EEE.EE..E.EE........E...EEE..EEE.EE.EEEE.EE..E...EE.
..EEE..E..E.E..EEE.EE....EE....EEE..E....E.EE.E.E....E...E.EEE....EE..EEE.E..EEEEEE.............E..EE...E....EEEE.E..E......E....EEE....E.....E..E.E...EE.EE.EEE....EEE.EEE..EEEE..E....EE..E.EE...E..E.
...E.E..EE.EEE..E..E..E....EEEE....EE.EEE....EEE..E..EE.E.EEE...E...E...E....E.EE.E.EE....EE..EEEE.E.E..E..EE..E..E.E.E.E...E.....EE..E.E...EE..E..E.E.EEEE...E.E.....EEEEEEEE....EE...E.E...E.E.E.EE.EE
.EE.E.E...EEEE.E.EE..E.....E.EE.E.E.E..EE.....EE.......E....EE....EEE....E.......E....EEEE.EE.E....E..EEE..E..EEEE..E...E...E..E.E.EEEE.E..EE.EE.......E....EEE.E.EE.E.EEEE.EEEEEE.E.EE...EE.......E.E..
EEE.E...E...E..EE.E.....EE.E....E.E.EE.....E.EE..EE..E..EEE...EE..E.E.E..E.EE.EE.E....E.......EE.EEEEEEEE..E.....E.E..E.E..E.......EE....EE.E...EE.E....EE...E.EE...EE.E.EE.E....EE.EE..E.E.E..EE.EEEE..
.E..E...E...E..E..E...EEE..EEE.EE..EE..E...E..E.E.....E...E.E.E.....E......E.EEE.E.EE.E.E.E.EE..EEEE.EE..EE.E..EEEEEE..EE..EE.EEEEE..EE.E.....EE.....E..E..EE.E.EE..E.......EE.EE....EEEEEE....EE.EEE...
..E.E.E....E.EEE.E.EEE....E.EEEEE..E..E.E.E.EEEEE..E.E...EE..EEE.......E.E..EEE.....EEE.E.E.E...EE.E.E.E...E....E.E.E.EE...EEE..EE...EE...EE.EE...E.EE.E.E..........EEEEE.E......E......EEEEEEEEE.E.EEEE
.EE..E....EE.....E.EE.EEEE.....EE....E.E........E..........E...EE..EE...EEEEE....EEE.E.EE.....E.EEEE.....EEE.E.E..E.EEE.E.EE.E....EE..E.E....EEEE.E.EE...E......E..E.E..E.EEE...EEEE....E.EE.E...E......
.E..E....E..EE...E...E...E..EEEE...E.E..E.EEE....E.E...E...EE...EEE.EEE.E.E.....EEEE...E......E.EEEE...E.......E..EEE..E..E....EE.E......E.E...EE.E...E...EE...EEEEE...E...E.E.E.EE.EE....E...EE......EE
.E.EEE.EEEE..E.E.EEEEE....E...E.E......E...EE.E..E.....E.E....E.EE.....E.E.E.EEE....E......E.E.E..EEE...E..E...E.E....E.E.....EE..EE.E..EEE.E.EE.E..........EE.E.E.E.E...E..EE..EE..E.EE.E.EEEEE..EE.EEE
.EEE....E.E....E...E.EE.E.EEE.E.....E....E.EEE.E.E..EEEEEEE..E...EEE..E.EEE.E.E.EE....E.E.E.EEE.E...E..EE..EEEE.EEE.E...EE.EE..E..E......EEEEEE.EEEEE.....E..E..E..EE....EE..E.E......EEEE.EE..EE.EE.EEE
E.EE.E.E.EEE..E...E.E..EEE....EEEEE...E..E....EE..EE...EEE....E.E...EE.EE....E.EE.EEEE.E.E.E...E.E..EE..EEEEE.E......E...EE.EEE.EE.EEEEE..E..EEE..E.EE..E..E..EE.EE..E.E..E..E.EEE.EE.E..EEEEE..EE.E.EE.
.E.E.E..EE..EEE.EE..E.EE........E...EEE..EEE.EE.EEEE.EE..E...EE...EEE..E..E.E..EEE.EE....EE....EEE..E....E.EE.E.E....E...E.EEE....EE..EEE.E..EEEEEE.............E..EE...E....EEEE.E..E......E....EEE....
E.....E..E.E...EE.EE.EEE....EEE.EEE..EEEE..E....EE..E.EE...E..E....E.E..EE.EEE..E..E..E....EEEE....EE.EEE....EEE..E..EE.E.EEE...E...E...E....E.EE.E.EE....EE..EEEE.E.E..E..EE..E..E.E.E.E...E.....EE..E.
E...EE..E..E.E.EEEE...E.E.....EEEEEEEE....EE...E.E...E.E.E.EE.EE.EE.E.E...EEEE.E.EE..E.....E.EE.E.E.E..EE.....EE.......E....EE....EEE....E.......E....EEEE.EE.E....E..EEE..E..EEEE..E...E...E..E.E.EEEE.
E..EE.EE.......E....EEE.E.EE.E.EEEE.EEEEEE.E.EE...EE.......E.E..EEE.E...E...E..EE.E.....EE.E....E.E.EE.....E.EE..EE..E..EEE...EE..E.E.E..E.EE.EE.E....E.......EE.EEEEEEEE..E.....E.E..E.E..E.......EE...
.EE.E...EE.E....EE...E.EE...EE.E.EE.E....EE.EE..E.E.E..EE.EEEE...E..E...E...E..E..E...EEE..EEE.EE..EE..E...E..E.E.....E...E.E.E.....E......E.EEE.E.EE.E.E.E.EE..EEEE.EE..EE.E..EEEEEE..EE..EE.EEEEE..EE.
E.....EE.....E..E..EE.E.EE..E.......EE.EE....EEEEEE....EE.EEE.....E.E.E....E.EEE.E.EEE....E.EEEEE..E..E.E.E.EEEEE..E.E...EE..EEEa
test cases: 4528, contracts: 3016+, errors: 1980, missing bytecode: 96, missing metadata: 96
```

### CODE REVIEW

**Comment by christianparpart** [comment]
For the output format, did you get inspired by ab (apache bench)? It reminds me of these times I used ab a lot. ;)

**Comment by cameel** [comment]
Maybe because there are so many errors :)

This format is pretty common. It's how Python's `unittest` prints results in non-verbose mode. We're also doing something kinda like this in `stopAfterParseTests.sh` (only dots though).

**Comment by cameel** [comment]
#10679 has been merged so this is now the next bytecode check PR to review.

File: test/scripts/test_bytecodecompare_prepare_report.py [hunk 1]
```diff
@@ -99,6 +99,58 @@ def test_format_report_should_not_print_anything_if_contract_report_list_is_empt
...
         self.assertEqual(report.format_report(), '')
 
+class TestPrepareReport_Statistics(unittest.TestCase):
+    def test_initialization(self):
+        self.assertEqual(Statistics(), Statistics(0, 0, 0, 0, 0))
+
+    def test_aggregate_bytecode_and_metadata_present(self):
+        statistics = Statistics()
+        statistics.aggregate(FileReport(file_name=Path('F'), contract_reports=[ContractReport('C', 'c.sol', 'B', 'M')]))
+        self.assertEqual(statistics, Statistics(1, 1, 0, 0, 0))
```
Comment (leonardoalt) [inline test/scripts/test_bytecodecompare_prepare_report.py:109]: What are these numbers?

Comment (leonardoalt) [reply test/scripts/test_bytecodecompare_prepare_report.py:109]: Ah just saw it on the other file

**Review by leonardoalt — APPROVED** [review]

File: test/scripts/test_bytecodecompare_prepare_report.py (hunk 1 above)
Comment (cameel) [reply test/scripts/test_bytecodecompare_prepare_report.py:109]: Expected values of the gathered statistics:
``` python
class Statistics:
    file_count: int = 0
    contract_count: int = 0
    error_count: int = 0
    missing_bytecode_count: int = 0
    missing_metadata_count: int = 0
```

**Pull request merged**